cclib/parser/gaussianparser.py

```diff
@@ -1853,46 +1857,86 @@ def natural_orbital_single_spin_parsing(inputfile, updateprogress_title):
+                    has_spin = 'spin' in line or 'Spin' in line
+                    has_charges = 'charges' in line or 'Charges' in line
+
+                    if has_charges and not hasattr(self, "atomcharges"):
+                        self.atomcharges = {}
+
+                    if has_spin and not hasattr(self, "atomspins"):
+                        self.atomspins = {}
+
+                    ones = next(inputfile)
+
+                    charges = []
+                    spins = []
+                    nline = next(inputfile)
 
-            elif "Lowdin" in line:
-                self.atomcharges["lowdin"] = charges
+                    # calculate how many lines need iterating over based on whether property is summed or not
+                    is_sum = 'summed' in line
+                    if is_sum:
+                        n = self.natom - self.nhydrogen
+                    else:
+                        n = self.natom
 
+                    # iterate over each line and append values to a list based on what property we have
+                    i = 0
+                    while i != n:
+                        if is_sum and nline.split()[1] == "H":  # some files included hydrogens in lists of summed charges with value 0.0, so these should not be recorded
+                            nline = next(inputfile)
+                        else:
+                            if has_charges:
+                                charges.append(float(nline.split()[2]))
+                            if has_spin and has_charges:
+                                spins.append(float(nline.split()[3]))
+                            if has_spin and not has_charges:
+                                spins.append(float(nline.split()[2]))
+                            if not i == n-1: # make sure we don't move to the header line for another set of values
+                                nline = next(inputfile)
+                            i +=1
+
+                    # input extracted values into self.atomcharges
+                    if prop in line:
+                        if has_charges:
+                            if is_sum:
+                                title_list = [header,"_charge_sum"]
+                                self.atomcharges["".join(title_list)] = charges
+                            else:
+                                title_list = [header,"_charge"]
+                                self.atomcharges["".join(title_list)] = charges
+                        if has_spin:
+                            if is_sum:
+                                title_list = [header,"_spin_sum"]
+                                self.atomspins["".join(title_list)] = spins
+                            else:
+                                title_list = [header,"_spin"]
+                                self.atomspins["".join(title_list)] = spins
+
+        if hasattr(self, "natom") and hasattr(self, "nhydrogen"):
+            extract_charges_spins("Mulliken","mul")
```

Comment:
here is another thing that I prefer doing although it is arguable. You wrote a function within the scope of a class function that modifies the class. That can be not very non-transparent to what is actually happening inside and what is being modified. My preference I guess would be to do the following but I am not 100% sure that is the right way:

- add the extract_charges_spins as a function at the same level of `extract` etc and pass in `self` as the first argument. Then it should probably work (your modifications of `self` work because `self` is passed in) as is, except maybe some variable that you are using and not passing in like `inputfile` but maybe that is a global variable anyway?

Reply:
I tried to follow the same convention they used a bit earlier in the file (line 1684) where they define a function within extract (def natural_orbital_single_spin_parsing). They then use that function just afterwards as part of extract, in the same sense as me with extract_charges_spins.